File: python/y2016/d02/aoc_2016_02.py
import sys
from common import timer, get_input


class KeyPad:
    def __init__(self, alt_layout=False):
        self.alt_layout = alt_layout
        if not self.alt_layout:
            self.keypad = [['1', '2', '3'], ['4', '5', '6'], ['7', '8', '9']]
        else:
            self.keypad = [['0', '0', '1', '0', '0'],
                           ['0', '2', '3', '4', '0'],
                           ['5', '6', '7', '8', '9'],
                           ['0', 'A', 'B', 'C', '0'],
                           ['0', '0', 'D', '0', '0']]

        self.directions = {
            "U": (-1, 0), "D": (1, 0), "L": (0, -1), "R": (0, 1)
        }
        self.mx = len(self.keypad[0]) - 1
        self.pos = [2, 0] if self.alt_layout else [1, 1]
        self.code = []

    def get_code(self, data):
        for line in data:
            self.code.append(self.keystring(line))

        try:
            return ''. join([d for d in self.code])
        except TypeError:
            print(self.code)

    def key(self, key):
        diff = self.directions[key]
        pos = [self.pos[0] + diff[0], self.pos[1] + diff[1]]

        if self.digit(pos):
            return pos
        else:
            return self.pos

    def digit(self, pos):
        if (0 <= pos[0] <= self.mx) and (0 <= pos[1] <= self.mx):
            digit = self.keypad[pos[0]][pos[1]]
            if digit != '0':
                return digit
        return 0

    def keystring(self, keystring):
        for c in keystring:
            self.pos = self.key(c)
        return self.digit(self.pos)


def clean_input(data):
    return data


@timer(part=1)
def solve_part_1(data):
    """Solution for part 1"""
    kp = KeyPad()

    return kp.get_code(data)


@timer(part=2)
def solve_part_2(data):
    """Solution for part 2"""
    kp = KeyPad(alt_layout=True)

    return kp.get_code(data)


@timer(part='main', title='Total elapsed', show_return=False)
def main(filename):
    data = clean_input(get_input(filename, 2016, 2))

    part1 = solve_part_1(data)
    part2 = solve_part_2(data)

    return part1, part2


if __name__ == "__main__":
    filename = "input.txt"
    if len(sys.argv) > 1:
        filename = sys.argv[1]
    main(filename)

File: C/2016/02/aoc_2016_02.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "aoc_types.h"
#include "aoc_utils.h"
#include "aoc_array.h"
#include "aoc_timer.h"
#include "aoc_string.h"

typedef struct {
    int x;
    int y;
} Directions_t;

typedef enum {
    KP_STD,
    KP_ALT,
    KP_COUNT
} KeyPadLayout;

typedef struct {
    KeyPadLayout layout;
    unsigned int rows;
    unsigned int columns;
    Directions_t directions[4];
    char **keys;
    char *code;
    unsigned int *pos;
} KeyPad_t;

#define NOT_FOUND -1
unsigned int command_get_index(char *commands, char command) {
    char needle[2];
    sprintf(needle, "%c", command);
    const char *p = strstr(commands, needle);
    if (p) {
        return (unsigned int)(p - commands);
    }
    return NOT_FOUND;
}

KeyPad_t *keypad_init(KeyPadLayout layout) {
    KeyPad_t *keypad = NULL;
    unsigned int i, j, index;
    char *keys[] = {"123456789", "0010002340567890ABC000D00"};

    keypad = (KeyPad_t *)calloc(1, sizeof(KeyPad_t));
    keypad->layout = layout;
    keypad->code = NULL;
    keypad->pos = (unsigned int *)calloc(2, sizeof(unsigned int));
    keypad->rows = 0;
    keypad->columns = 0;
    if (layout == KP_STD) {
        keypad->rows = 3;
        keypad->columns = 3;
        keypad->pos[0] = 1;
        keypad->pos[1] = 1;
    } else {
        keypad->rows= 5;
        keypad->columns = 5;
        keypad->pos[0] = 0;
        keypad->pos[1] = 2;
    }

    keypad->keys = (char **)malloc(sizeof(char *) * keypad->rows);
    for (i = 0; i < keypad->rows; i++) {
        keypad->keys[i] = (char *)calloc(keypad->columns, sizeof(char));
        for (j = 0; j < keypad->columns; j++) {
            index = (j + i*keypad->columns);
            keypad->keys[i][j] = keys[layout][index];
        }
    }

    keypad->directions[0].x = 0;
    keypad->directions[0].y = -1;
    keypad->directions[1].x = 0;
    keypad->directions[1].y = 1;
    keypad->directions[2].x = -1;
    keypad->directions[2].y = 0;
    keypad->directions[3].x = 1;
    keypad->directions[3].y = 0;

    return keypad;
}

void keypad_destroy(KeyPad_t *keypad) {
    for (unsigned int i = 0; i < keypad->rows; i++) {
        free(keypad->keys[i]);
    }
    free(keypad->keys);
    free(keypad->code);
    free(keypad->pos);
    free(keypad);
}

char keypad_digit(KeyPad_t *keypad, unsigned int *pos) {
    char digit;
    if ((pos[0] <= keypad->columns - 1) &&
        (pos[1] <= keypad->rows - 1)) {
        digit = keypad->keys[pos[1]][pos[0]];
        if (digit != '0') {
            return digit;
        }
    }
    return 0;
}

unsigned int *keypad_key(KeyPad_t *keypad, char key) {
    Directions_t diff;
    unsigned int pos[2];
    unsigned int index;

    index = command_get_index("UDLR", key);

    diff = keypad->directions[index];
    pos[0] = keypad->pos[0] + diff.x;
    pos[1] = keypad->pos[1] + diff.y;

    if ((keypad_digit(keypad, pos))) {
        keypad->pos[0] = pos[0];
        keypad->pos[1] = pos[1];
        return keypad->pos;
    }
    return keypad->pos;
}

char keypad_keystring(KeyPad_t *keypad, char *line) {
    unsigned long i;

    for (i = 0; i < strlen(line); i++) {
        keypad->pos = keypad_key(keypad, line[i]);
    }
    return keypad_digit(keypad, keypad->pos);
}

char *keypad_get_code(KeyPad_t *keypad, AocArrayPtr data) {
    char *line = NULL;
    unsigned int i;
    unsigned int data_length;

    data_length = aoc_array_length(data);
    keypad->code = (char *)calloc(data_length + 1, sizeof(char));
    for (i = 0; i < aoc_array_length(data); i++) {
        line = aoc_str_array_index(data, i);
        keypad->code[i] = keypad_keystring(keypad, line);
    }
    return keypad->code;
}

void *solve_part_1(AocData_t *data) {
    KeyPad_t *keypad = NULL;
    char *return_string;

    keypad = keypad_init(KP_STD);
    return_string = strdup(keypad_get_code(keypad, data->data));
    keypad_destroy(keypad);
    return return_string;
}

void *solve_part_2(AocData_t *data) {
    KeyPad_t *keypad = NULL;
    char *return_string;

    keypad = keypad_init(KP_ALT);
    return_string = strdup(keypad_get_code(keypad, data->data));
    keypad_destroy(keypad);
    return return_string;
}

void *solve_all(AocData_t *data) {

    if (data->data) {
        timer_func(1, solve_part_1, data, 1);
        timer_func(2, solve_part_2, data, 1);
    }

    return NULL;
}

int main(int argc, char **argv) {
    AocData_t *data;

    char sourcefile[20];
    int year, day;

    strcpy(sourcefile, aoc_basename(__FILE__));
    sscanf(sourcefile, "aoc_%4d_%02d.c", &year, &day);

    if (argc > 1) {
        data = aoc_data_new(argv[1], year, day);
    } else {
        data = aoc_data_new("input.txt", year, day);
    }

    printf("================================================\n");
    printf("Solution for %d, day %02d\n", year, day);
    timer_func(0, solve_all, data, 0);

    aoc_data_free(data);

    return 0;
}
